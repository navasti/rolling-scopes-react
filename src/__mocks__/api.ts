import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  PokemonMove,
  PokemonType,
  Pokemon,
} from 'types';
import { hasError } from 'utils';

export const fetchBaseMock = (mockData: Pokemon | PokemonMove | PokemonType) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = (
  mockData: PokemonDetails | PokemonMoveDetails | PokemonTypeDetails
) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const handleResponseMock = (status: number, mockData: Pokemon) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};

export const handleMappedResponseMock = (status: number, mockData: PokemonDetails) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};
