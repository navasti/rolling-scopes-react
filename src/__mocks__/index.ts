export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
