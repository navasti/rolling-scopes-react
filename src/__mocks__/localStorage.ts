export const emptyLocalStorageMock = (() => {
  let store: Record<string, string> = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();

export const localStorageMock = (key: string, initialValue: string) => {
  let store: Record<string, string> = {};
  store[key] = initialValue;
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
};
