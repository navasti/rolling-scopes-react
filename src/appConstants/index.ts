import { FormFields } from 'types';
import * as yup from 'yup';

export enum Limits {
  pokemon = 20,
  type = 10,
  move = 15,
}

export enum ErrorStatuses {
  success = '2',
  client = '4',
  server = '5',
}

export enum AvailableTabs {
  pokemons = 'pokemons',
  moves = 'moves',
  types = 'types',
}

export enum Fields {
  birthday = 'birthday',
  consent = 'consent',
  gender = 'gender',
  avatar = 'avatar',
  shiny = 'shiny',
  name = 'name',
  type = 'type',
}

export enum ErrorMessages {
  name = 'Name must be at least 2 characters long.',
  consent = 'Regulations consent is required.',
  birthday = 'Birthday is required.',
  type = 'Main type is required.',
  gender = 'Gender is required.',
}

export enum Types {
  fire = 'fire',
  grass = 'grass',
  poison = 'poison',
  flying = 'flying',
  ground = 'ground',
  psycho = 'psycho',
  water = 'water',
  ice = 'ice',
}

export const NONE = 'none';
export const MALE = 'male';
export const FEMALE = 'female';
export const INPUT_VALUE_KEY = 'input-value';
export const SUCCESS_MESSAGE = 'Pokemon added successfuly';
export const LINKS = ['home', 'about', 'form', 'invalid', 'details'] as const;
export const SEARCH_BAR_INSTRUCTIONS = [
  'Type and press enter to search for specific pokemon, type or move.',
  'Clear input and press enter to search for all pokemons, types and moves.',
] as const;

export const TABS = Object.values(AvailableTabs);
export const POKEMON_TYPES = Object.values(Types);

export const FORM_VALIDATION_SCHEMA = yup.object().shape({
  type: yup.string().required(ErrorMessages.type),
  name: yup.string().min(2, ErrorMessages.name).required(ErrorMessages.name),
  gender: yup.string().nullable().required(ErrorMessages.gender),
  avatar: yup
    .mixed()
    .test('fileSize', 'The file is too large', (value) =>
      !value?.length ? true : value?.[0]?.size <= 200000
    ),
  birthday: yup.string().required(ErrorMessages.birthday),
  consent: yup.bool().oneOf([true], ErrorMessages.consent),
  shiny: yup.bool(),
});

const BASE = 'https://pokeapi.co/api/v2';

export const API = {
  POKEMON_LIMIT: `?limit=${Limits.pokemon}`,
  TYPE_LIMIT: `?limit=${Limits.type}`,
  MOVE_LIMIT: `?limit=${Limits.move}`,
  POKEMON: `${BASE}/pokemon`,
  TYPE: `${BASE}/type`,
  MOVE: `${BASE}/move`,
  BASE,
};

export const FIELDS_VALIDATION_BY_NAME = {
  TEXT: [Fields.name, Fields.birthday, Fields.type],
  CHECK: [Fields.consent, Fields.gender],
};

export const DEFAULT_VALUES: FormFields = {
  consent: false,
  avatar: null,
  gender: MALE,
  shiny: false,
  birthday: '',
  name: '',
  type: '',
};
