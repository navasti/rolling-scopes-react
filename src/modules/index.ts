export { LocationBadge } from './Header/components';
export { DetailsModal } from './DetailsModal';
export { SearchPage } from './SearchPage';
export { NotFound } from './NotFound';
export { Header } from './Header';
export { Layout } from './Layout';
export { About } from './About';
export { Form } from './Form';
export { App } from './App';
