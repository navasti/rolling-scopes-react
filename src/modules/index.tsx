export { App } from './App';
export { Form } from './Form';
export { About } from './About';
export { Layout } from './Layout';
export { Header } from './Header';
export { NotFound } from './NotFound';
export { SearchPage } from './SearchPage';
export { LocationBadge } from './Header/components';
export { Card, SearchBar } from './SearchPage/components';
