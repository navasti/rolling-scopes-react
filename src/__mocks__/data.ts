import { ErrorMessages } from 'appConstants';
import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  PokemonData,
  PokemonMove,
  PokemonType,
  Details,
  Pokemon,
  Lengths,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const links = ['Home', 'About', 'Form', 'Invalid'];

export const testingComponentName = 'Test';
export const testingLocation = '/test';

export const testingContent = 'Content';
export const testingTitle = 'Title';

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

//

export const lengths: Lengths = { pokemons: 10, moves: 15, types: 20 };

const samplePokemon = (): PokemonDetails => ({
  base_experience: 62,
  name: 'charmander',
  weight: 85,
  height: 6,
  id: 4,
  abilities: [
    {
      ability: { name: 'blaze', url: 'https://pokeapi.co/api/v2/ability/66/' },
      is_hidden: false,
      slot: 1,
    },
  ],
  sprites: {
    front_default: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
    back_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/4.png',
  },
  stats: [
    {
      base_stat: 52,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
  ],
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'https://pokeapi.co/api/v2/type/10/',
      },
    },
  ],
});

const sampleMove = (): PokemonMoveDetails => ({
  accuracy: 100,
  name: 'tackle',
  priority: 0,
  power: 40,
  id: 33,
  pp: 35,
  learned_by_pokemon: [
    {
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
    },
  ],
  type: {
    name: 'normal',
    url: 'https://pokeapi.co/api/v2/type/1/',
  },
});

const sampleType = (): PokemonTypeDetails => ({
  moves: [{ name: 'gust', url: 'https://pokeapi.co/api/v2/move/16/' }],
  name: 'flying',
  id: 3,
  damage_relations: {
    double_damage_from: [
      { name: 'electric', url: 'https://pokeapi.co/api/v2/type/13/' },
      { name: 'rock', url: 'https://pokeapi.co/api/v2/type/6/' },
    ],
    double_damage_to: [{ name: 'fighting', url: 'https://pokeapi.co/api/v2/type/2/' }],
    half_damage_from: [{ name: 'bug', url: 'https://pokeapi.co/api/v2/type/7/' }],
    half_damage_to: [{ name: 'steel', url: 'https://pokeapi.co/api/v2/type/9/' }],
    no_damage_from: [{ name: 'ground', url: 'https://pokeapi.co/api/v2/type/5/' }],
    no_damage_to: [],
  },
  pokemon: [
    { pokemon: { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' }, slot: 2 },
  ],
});

const prepareBuilder = <T extends Details>(sample: () => T) => {
  const builder = (object = sample()) => ({
    withName: (name: T['name']) => builder({ ...object, name }),
    withId: (id: T['id']) => builder({ ...object, id }),
    valueOf: () => object,
  });
  return builder;
};

const pokemonBuilder = prepareBuilder(samplePokemon);
const moveBuilder = prepareBuilder(sampleMove);
const typeBuilder = prepareBuilder(sampleType);

export const pokemonsMock: Array<PokemonDetails> = [
  pokemonBuilder().withId(1).withName('bulbasaur'),
  pokemonBuilder().withId(25).withName('pikachu'),
  pokemonBuilder(),
].map((builder) => builder.valueOf());

export const movesMock: Array<PokemonMoveDetails> = [
  moveBuilder().withId(55).withName('water-gun'),
  moveBuilder().withId(50).withName('disable'),
  moveBuilder(),
].map((builder) => builder.valueOf());

export const typesMock: Array<PokemonTypeDetails> = [
  typeBuilder().withId(2).withName('fighting'),
  typeBuilder().withId(5).withName('ground'),
  typeBuilder(),
].map((builder) => builder.valueOf());
