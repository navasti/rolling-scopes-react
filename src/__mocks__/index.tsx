import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
