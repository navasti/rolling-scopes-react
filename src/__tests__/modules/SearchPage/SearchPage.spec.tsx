import { screen, render, waitForElementToBeRemoved, act } from '@testing-library/react';
import { AvailableTabs, INPUT_VALUE_KEY, TABS } from 'appConstants';
import { ChangeEvent, KeyboardEvent } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { SearchPage } from 'modules';
import {
  Lengths,
  Pokemon,
  PokemonMove,
  PokemonType,
  PokemonDetails,
  PokemonMoveDetails,
  PokemonTypeDetails,
} from 'types';
import {
  typesMock,
  movesMock,
  pokemonsMock,
  testingLocation,
  testingComponentName,
} from '__mocks__/data';
import { localStorageMock } from '__mocks__/localStorage';

type CardsProps = {
  types: Array<PokemonTypeDetails>;
  moves: Array<PokemonMoveDetails>;
  pokemons: Array<PokemonDetails>;
  activeTab: AvailableTabs;
  isLoading: boolean;
};

type LayoutProps = {
  children: JSX.Element;
  componentName: string;
  location: string;
};

type TabsProps = {
  onClick: (tab: AvailableTabs) => void;
  activeTab: AvailableTabs;
  options: typeof TABS;
  isLoading: boolean;
  lengths: Lengths;
};

type SearchBarProps = {
  onKeyDown: (event: KeyboardEvent<HTMLInputElement>) => void;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  inputValue: string;
  isLoading: boolean;
  label: string;
};

const localStorage = localStorageMock(INPUT_VALUE_KEY, 'testing-value');

Object.defineProperty(window, 'localStorage', {
  value: localStorage,
});

jest.mock('modules', () => {
  const { SearchPage } = jest.requireActual('modules/SearchPage');
  return {
    __esModule: true,
    SearchPage,
    Layout: ({ componentName, children, location }: LayoutProps) => (
      <div>
        <p>
          <span data-testid="component-mock">{componentName}</span>
          <span data-testid="location-mock">{location}</span>
        </p>
        <main>{children}</main>
      </div>
    ),
  };
});

jest.mock('modules/SearchPage/components', () => {
  return {
    __esModule: true,
    SearchBar: ({ onChange, onKeyDown }: SearchBarProps) => {
      return (
        <div>
          <div>
            <label htmlFor="input-mock">testing label</label>
            <svg />
            <input
              data-testid="input-mock"
              value="testing-value"
              onKeyDown={onKeyDown}
              onChange={onChange}
              id="input-mock"
              type="text"
            />
          </div>
          <div>
            <span>Type and press enter to search for specific pokemon, type or move.</span>
            <span>Clear input and press enter to search for all pokemons, types and moves.</span>
          </div>
        </div>
      );
    },
    Tabs: ({ isLoading, lengths }: TabsProps) => (
      <div>
        <span data-testid="pokemons-length-mock">
          pokemons {isLoading ? null : `(${lengths.moves})`}
        </span>
        <span data-testid="moves-length-mock">moves {isLoading ? null : `(${lengths.moves})`}</span>
        <span data-testid="types-length-mock">types {isLoading ? null : `(${lengths.types})`}</span>
      </div>
    ),
    Cards: (props: CardsProps) => {
      return (
        <div data-testid="cards-mock">
          {props.isLoading ? (
            <p>loading</p>
          ) : (
            <div>
              <p>{props.types[0].name}</p>
              <p>{props.moves[0].name}</p>
              <p>{props.pokemons[0].name}</p>
            </div>
          )}
        </div>
      );
    },
  };
});

jest.mock('utils/api.ts', () => {
  const { pokemonsMock, movesMock, typesMock } = require('__mocks__/data');
  return {
    _esModule: true,
    fetchTypeByParameter: async () => Promise.resolve(typesMock[0]),
    fetchMoveByParameter: async () => Promise.resolve(movesMock[0]),
    fetchPokemonByParameter: async () => Promise.resolve(pokemonsMock[0]),
    fetchPokemonBase: async (): Promise<Array<Pokemon>> => Promise.resolve([]),
    fetchMoveBase: async (): Promise<Array<PokemonMove>> => Promise.resolve([]),
    fetchTypeBase: async (): Promise<Array<PokemonType>> => Promise.resolve([]),
    fetchTypeDetails: async (): Promise<PokemonTypeDetails[]> => Promise.resolve(typesMock),
    fetchMoveDetails: async (): Promise<PokemonMoveDetails[]> => Promise.resolve(movesMock),
    fetchPokemonDetails: async (): Promise<PokemonDetails[]> => Promise.resolve(pokemonsMock),
  };
});

describe('SearchPage', () => {
  it('Layout should be rendered with componentName and location props given further', () => {
    act(() => {
      render(<SearchPage componentName={testingComponentName} location={testingLocation} />, {
        wrapper: BrowserRouter,
      });
    });
    expect(screen.getByTestId('component-mock')).toHaveTextContent(testingComponentName);
    expect(screen.getByTestId('location-mock')).toHaveTextContent(testingLocation);
  });
  it('Tabs should be rendered and display length of data after fetching', async () => {
    act(() => {
      render(<SearchPage componentName={testingComponentName} location={testingLocation} />, {
        wrapper: BrowserRouter,
      });
    });
    expect(screen.getByTestId('pokemons-length-mock')).toHaveTextContent('pokemons');
    expect(screen.getByTestId('moves-length-mock')).toHaveTextContent('moves');
    expect(screen.getByTestId('types-length-mock')).toHaveTextContent('types');
    expect(screen.queryByText('loading')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.queryAllByText('loading'));
    expect(screen.getByTestId('pokemons-length-mock')).toHaveTextContent('pokemons (1)');
    expect(screen.getByTestId('moves-length-mock')).toHaveTextContent('moves (1)');
    expect(screen.getByTestId('types-length-mock')).toHaveTextContent('types (1)');
    expect(screen.getByText(pokemonsMock[0].name)).toBeVisible();
    expect(screen.getByText(movesMock[0].name)).toBeVisible();
    expect(screen.getByText(typesMock[0].name)).toBeVisible();
  });
  it('SearchBar should be rendered and localStorage input should be visible', () => {
    act(() => {
      render(<SearchPage componentName={testingComponentName} location={testingLocation} />, {
        wrapper: BrowserRouter,
      });
    });
    const input = screen.getByLabelText('testing label');
    const instructions = screen.getAllByText('and press enter', { exact: false });
    instructions.forEach((instruction) => expect(instruction).toBeVisible());
    expect(instructions).toHaveLength(2);
    expect(input).toBeVisible();
  });
});
