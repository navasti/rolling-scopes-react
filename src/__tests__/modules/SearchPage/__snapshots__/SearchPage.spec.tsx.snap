// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should match snapshot before and after data fetching 1`] = `
<div>
  <div>
    <p>
      <span
        data-testid="component-mock"
      >
        Test
      </span>
      <span
        data-testid="location-mock"
      >
        /test
      </span>
    </p>
    <main>
      <div
        class="sc-bczRLJ frUuCi"
      >
        <div>
          <div>
            <label
              for="input-mock"
            >
              testing label
            </label>
            <svg />
            <input
              data-testid="input-mock"
              id="input-mock"
              type="text"
              value="testing-value"
            />
          </div>
          <div>
            <span>
              Type and press enter to search for specific pokemon, type or move.
            </span>
            <span>
              Clear input and press enter to search for all pokemons, types and moves.
            </span>
          </div>
        </div>
        <div>
          <span
            data-testid="pokemons-length-mock"
          >
            pokemons 
          </span>
          <span
            data-testid="moves-length-mock"
          >
            moves 
          </span>
          <span
            data-testid="types-length-mock"
          >
            types 
          </span>
        </div>
        <div
          data-testid="cards-mock"
        >
          <p>
            loading
          </p>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`SearchPage should match snapshot before and after data fetching 2`] = `
<div>
  <div>
    <p>
      <span
        data-testid="component-mock"
      >
        Test
      </span>
      <span
        data-testid="location-mock"
      >
        /test
      </span>
    </p>
    <main>
      <div
        class="sc-bczRLJ frUuCi"
      >
        <div>
          <div>
            <label
              for="input-mock"
            >
              testing label
            </label>
            <svg />
            <input
              data-testid="input-mock"
              id="input-mock"
              type="text"
              value="testing-value"
            />
          </div>
          <div>
            <span>
              Type and press enter to search for specific pokemon, type or move.
            </span>
            <span>
              Clear input and press enter to search for all pokemons, types and moves.
            </span>
          </div>
        </div>
        <div>
          <span
            data-testid="pokemons-length-mock"
          >
            pokemons 
            (1)
          </span>
          <span
            data-testid="moves-length-mock"
          >
            moves 
            (1)
          </span>
          <span
            data-testid="types-length-mock"
          >
            types 
            (1)
          </span>
        </div>
        <div
          data-testid="cards-mock"
        >
          <div>
            <p>
              fighting
            </p>
            <p>
              water-gun
            </p>
            <p>
              bulbasaur
            </p>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
